internal/contour/snapshot.go

```diff
@@ -40,7 +52,70 @@ func NewSnapshotHandler(c cache.SnapshotCache, resources []ResourceCache, logger
 	}
 }
 
-// OnChange is called when the DAG is rebuilt
-// and a new snapshot is needed.
+// OnChange is called when the DAG is rebuilt and a new snapshot is needed.
 func (s *SnapshotHandler) OnChange(root *dag.DAG) {
+
+	// Generate new snapshot version.
+	snapshotVersion := s.newSnapshotVersion()
+
+	// Create an snapshot with all xDS resources.
+	snapshot := cache.Snapshot{}
+
+	// Loop through all the xDS resources and pass them to the snapshot
+	for _, r := range s.resources {
+		snapshot.Resources[responseTypeOf(r.TypeURL())] = cache.NewResources(snapshotVersion, asResources(r.Contents()))
```

Comment:
Threw me for a second that `snapshot.Resources` is an array, not a map or even a slice, and that the response types being used to index are ints. Might be worth a comment somewhere.